ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -98,17 +105,59 @@ private[platform] case class StateCache[K, V](
     registerUpdateTimer,
     pendingUpdates.synchronized {
       val validAt = cacheIndex
-      val eventualValue = Future.delegate(fetchAsync(validAt))
-      val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdatesState.empty)
-      if (pendingUpdatesForKey.latestValidAt < validAt) {
-        pendingUpdatesForKey.latestValidAt = validAt
-        pendingUpdatesForKey.pendingCount += 1
-        registerEventualCacheUpdate(key, eventualValue, validAt)
-          .flatMap(_ => eventualValue)
-      } else eventualValue
+      putPendingForKey(key, validAt, Future.delegate(fetchAsync(validAt)))
     },
   )
 
+  private def putPendingForKey(key: K, validAt: Offset, eventualValue: Future[V])(implicit
+      loggingContext: LoggingContext
+  ): Future[V] = {
+    val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdatesState.empty)
+    if (pendingUpdatesForKey.latestValidAt < validAt) {
+      pendingUpdatesForKey.latestValidAt = validAt
+      pendingUpdatesForKey.pendingCount += 1
+      registerEventualCacheUpdate(key, eventualValue, validAt)
+        .flatMap(_ => eventualValue)
+    } else eventualValue
+  }
+
+  /** Batch fetches many contracts at once to avoid incremental fetches to database */
+  def putAsyncMany(keys: Seq[K], fetchAsync: Offset => Future[Map[K, V]], empty: K => Future[V])(
+      implicit loggingContext: LoggingContext
+  ): Future[Unit] =
+    Future
+      .sequence(
```

Comment:
probably okay, but the result future will complete with failure immediately after the first failure, and not waiting for the other futures to finish